File: java/06_018.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * A basic binary tree node implementation
 */
class TreeNode {
   int value;
   TreeNode left;
   TreeNode right;
   
   /**
    * Create tree node with given value
    */
   public TreeNode(int value) {
      this.value = value;
      this.left = null;
      this.right = null;
   }
}

/**
 * Depth-First Search (DFS) tree traversal
 */
class DFSTree {
   /**
    * Recursive DFS traversal in pre-order
    */
   public List<Integer> dfsRecursive(TreeNode root) {
      List<Integer> result = new ArrayList<>();
      explore(root, result);
      return result;
   }
   
   /**
    * Helper method for recursive traversal
    */
   private void explore(TreeNode node, List<Integer> result) {
      if (node == null) {
         return;
      }
      
      // Pre-order: Process node first
      result.add(node.value);
      
      // Then go left and right
      explore(node.left, result);
      explore(node.right, result);
   }
   
   /**
    * Iterative DFS traversal using stack
    */
   public List<Integer> dfsIterative(TreeNode root) {
      if (root == null) {
         return new ArrayList<>();
      }
      
      List<Integer> result = new ArrayList<>();
      Stack<TreeNode> stack = new Stack<>();
      stack.push(root);
      
      while (!stack.isEmpty()) {
         // Get and process top node
         TreeNode node = stack.pop();
         result.add(node.value);
         
         // Push right first so left is processed first
         if (node.right != null) {
            stack.push(node.right);
         }
         if (node.left != null) {
            stack.push(node.left);
         }
      }
      
      return result;
   }
}

File: python/06_018.py
class TreeNode:
   """Binary tree node."""
   def __init__(self, value):
      self.value = value
      self.left = None
      self.right = None

class DFSTree:
   """Depth-First Search tree traversal."""
   def dfs_recursive(self, root):
      """
      Recursive pre-order DFS traversal.
      
      Time Complexity: O(n)
      Space Complexity: O(h)
      """
      result = []
      
      def explore(node):
         if not node:
            return
         
         # Pre-order: root, left, right
         result.append(node.value)
         explore(node.left)
         explore(node.right)
      
      explore(root)
      return result
   
   def dfs_iterative(self, root):
      """
      Iterative pre-order DFS traversal.
      
      Time Complexity: O(n)
      Space Complexity: O(h)
      """
      if not root:
         return []
      
      result = []
      stack = [root]
      
      while stack:
         node = stack.pop()
         result.append(node.value)
         
         # Right first, then left (stack order)
         if node.right:
            stack.append(node.right)
         if node.left:
            stack.append(node.left)
      
      return result

# Demonstrate DFS traversals
if __name__ == "__main__":
   # Create sample tree
   #        1
   #      /   \
   #     2     3
   #    / \   / \
   #   4   5 6   7
   root = TreeNode(1)
   root.left = TreeNode(2)
   root.right = TreeNode(3)
   root.left.left = TreeNode(4)
   root.left.right = TreeNode(5)
   root.right.left = TreeNode(6)
   root.right.right = TreeNode(7)
   
   # Initialize DFS traversal
   dfs = DFSTree()
   
   # Test recursive and iterative DFS
   print("Recursive DFS:", dfs.dfs_recursive(root))
   print("Iterative DFS:", dfs.dfs_iterative(root))